### PR TITLE
[ISSUE #4598] Improve eventmesh-connector-lark

### LINKED ISSUE #4598: [Enhancement] Improve eventmesh-connector-lark
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/eventmesh/issues?q=is%3Aissue) and found no similar issues.

### Enhancement Request

There are still many optimization aspects of eventmesh-connectot-feishu module.

### Describe the solution you'd like

1. fix https://github.com/apache/eventmesh/pull/4522#discussion_r1411710473
2. Rename the module to eventmesh-connector-lark
3. Simple http calls are changed to use the sdk provided by lark
4. Support sinking text and markdown text types
5. Support `@` user or everyone
6. Provide documentation
7. Prevent message retransmission
---

1. 修复 https://github.com/apache/eventmesh/pull/4522#discussion_r1411710473
2. 将模块重命名为eventmesh-connector-lark
3. 简单的http调用改为使用lark提供的sdk
4. 支持下沉text和markdown的文本类型
5. 支持`@`用户或者所有人
6. 提供文档
8. 防止消息重传

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct) *

### PR DESCRIPTION
**Pull request by hhuang1231**
Fixes #4598 .

### Motivation

Improve eventmesh-connector-feishu.

### Modifications

See #4598  for details.

### Output

image

### Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
- If a feature is not applicable for documentation, explain why?
- If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: eventmesh-connectors/eventmesh-connector-lark/src/main/resources/sink-config.yml [hunk 1]
```diff
@@ -17,14 +17,18 @@
...
-  reciveId: reciveIdValue
-  reciveType: open_id
+  userName: larkSinkUser
+  passWord: larkPassWord
+sinkConnectorConfig:
+  connectorName: larkSink
+  appId: appId
+  appSecret: appSecret
+  receiveIdType: open_id
+  receiveId: receiveId
```
[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/main/resources/sink-config.yml:32]: In addition to configuration reading problems, the original Feishu Sink also had CouldEvent.Data reading problems:
原先的Feishu Sink除了配置读取问题，还有CouldEvent.Data的读取问题：
image

The following message will probably be generated:
大概会生成如下的消息:
image

File: eventmesh-connectors/eventmesh-connector-lark/build.gradle [hunk 2]
```diff
@@ -15,19 +15,20 @@
...
 
 dependencies {
     api project(":eventmesh-openconnect:eventmesh-openconnect-java")
-    implementation feishu
+    implementation lark
     compileOnly 'org.projectlombok:lombok'
     annotationProcessor 'org.projectlombok:lombok'
 
-    testImplementation "org.mockito:mockito-core"
+    //testImplementation "org.mockito:mockito-core"
```
[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/build.gradle:31]: I will remove it.

File: eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/ImServiceHandler.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+import com.lark.oapi.service.im.v1.model.CreateMessageResp;
+import com.lark.oapi.service.im.v1.model.ext.MessageText;
+
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ImServiceHandler {
+
+    private static final ConcurrentHashMap<ConnectRecord, CreateMessageReq> UN_ACK_REQ = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/ImServiceHandler.java:71]: - The main function of this collection is to track requests that do not return a response. 
- The interface exposed by the Lark platform can guarantee us that if the UUID is the same, the message will not be sent repeatedly. A UUID will be injected inside `CreateMessageReq`. 
- The original code will re-inject a new UUID every time it is retransmitted, which will lead to repeated consumption by Feishu.

- 这个集合的主要作用是追踪那些未返回响应的请求。
- 飞书平台暴露的接口可以向我们保证——如果UUID一样，那么消息就不会重复发送。
- 在`CreateMessageReq`内部将会注入UUID，原先的代码会在每次重传时，都重新注入一个新的UUID，这将会导致飞书重复消费。

File: eventmesh-connectors/eventmesh-connector-lark/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# eventmesh-connector-lark
```
[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/README.md:1]: ok

File: eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/ConnectRecordExtensionKeys.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ */
+public interface ConnectRecordExtensionKeys {
+
+    /**
+     * {@code text} or {@code markdown}, otherwise use {@code text} to replace.
+     */
+    String TEMPLATE_TYPE_4_LARK = "templatetype4lark";
+
+    /**
+     * The value format is {@code id,name;id,name;}. <br/>
```
[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/ConnectRecordExtensionKeys.java:31]: I will replace them by `<p>`

**Comment by xwm1992** [comment]
I have question about why you change the module name from Feishu to lark ?

**Comment by hhuang1231** [comment]
> I have question about why you change the module name from Feishu to lark ?

I found that the dingding module was changed to dingtalk.So I think it is necessary to change Feishu to lark.

File: eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/connector/LarkSinkConnector.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            try {
+                imServiceHandler.sink(connectRecord);
+            } catch (ExecutionException | RetryException e) {
+                log.error("Failed to sink event to lark", e);
+            }
+        }
+    }
+
+    @SneakyThrows
+    public static String getTenantAccessToken(String appId, String appSecret) {
```
[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/connector/LarkSinkConnector.java:125]: Lark's access credentials have an expiration mechanism, but the expiration time can be refreshed by re-obtaining new credentials within half an hour before expiration. The documentation provided by Lark is as follows:
> The maximum validity period of tenant_access_token is 2 hours. If this interface is called when the validity period is less than 30 minutes, a new tenant_access_token will be returned, and there will be two valid tenant_access_tokens at the same time.

_I think expiration is not the reason for using Cache. If you only need to manage one kind of credential, then there is no need to use a collection to maintain the validity of the credential._

---

飞书的访问凭证存在过期机制，但是在过期前半个小时内重新获取新的凭证即可刷新过期时间。飞书提供的文档说明如下：
>  tenant_access_token 的最大有效期是 2 小时。如果在有效期小于 30 分钟的情况下，调用本接口，会返回一个新的 tenant_access_token，这会同时存在两个有效的 tenant_access_token。

_我认为过期并不是使用Cache的原因，如果仅需要管理一种凭证，那么并不需要使用一个集合来维护凭证的有效性。_

[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/connector/LarkSinkConnector.java:125]: 1. For API documentation, see [here](https://open.larksuite.com/document/server-docs/getting-started/api-access-token/auth-v3/tenant_access_token_internal)
2. The following code will use the sdk provided by lark to request access credentials.
    ```java
     TenantAccessTokenResp resp = client.ext().getTenantAccessTokenBySelfBuiltApp(
                     SelfBuiltTenantAccessTokenReq.newBuilder()
                             .appSecret(appSecret)
                             .appId(appId)
                             .build());

3. I don’t understand what reliability means. I hope you can give me a more detailed answer.

---
1. API文档参见[这里](https://open.feishu.cn/document/server-docs/authentication-management/access-token/tenant_access_token_internal)
2. 下面这段代码将使用飞书提供的sdk请求获取访问凭证
   ```java
    TenantAccessTokenResp resp = client.ext().getTenantAccessTokenBySelfBuiltApp(
                    SelfBuiltTenantAccessTokenReq.newBuilder()
                            .appSecret(appSecret)
                            .appId(appId)
                            .build());

3. 我不明白可靠性是指什么，希望您能给我更详细的解答。

File: eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/ImServiceHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/ImServiceHandler.java:71]: ok, I will continue to implement asynchronous requests in this PR according to your suggestions.

**Comment by hhuang1231** [comment]
@pandaapo PLAT, thank you!

File: eventmesh-connectors/eventmesh-connector-lark/src/test/java/org/apache/eventmesh/connector/lark/sink/LarkSinkConnectorTest.java [hunk 7]
```diff
@@ -43,12 +43,17 @@
 import org.mockito.Mock;
 import org.mockito.MockedStatic;
 import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.junit.jupiter.MockitoSettings;
+import org.mockito.quality.Strictness;
 
 @ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
```
[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/test/java/org/apache/eventmesh/connector/lark/sink/LarkSinkConnectorTest.java:50]: If this annotation is not added, the following exception message will be thrown
```java
org.mockito.exceptions.misusing.UnnecessaryStubbingException: 
Unnecessary stubbings detected.
Clean & maintainable test code requires zero unnecessary code.
Following stubbings are unnecessary (click to navigate to relevant line of code):
  1. -> at org.apache.eventmesh.connector.lark.sink.LarkSinkConnectorTest.setup(LarkSinkConnectorTest.java:68)
Please remove unnecessary stubbings or use 'lenient' strictness. More info: javadoc for UnnecessaryStubbingException class.

File: eventmesh-connectors/eventmesh-connector-lark/src/test/java/org/apache/eventmesh/connector/lark/sink/ImServiceHandlerTest.java [hunk 8]
```diff
@@ -85,31 +88,75 @@ public void setup() throws Exception {
...
             ConnectRecord connectRecord = new ConnectRecord(partition, offset,
                     System.currentTimeMillis(), "test-lark".getBytes(StandardCharsets.UTF_8));
-            Assertions.assertThrows(Exception.class, () -> imServiceHandler.sink(connectRecord));
-        }
+            if (Boolean.parseBoolean(sinkConnectorConfig.getSinkAsync())) {
+                imServiceHandler.sinkAsync(connectRecord);
 
-        // (maxRetryTimes + 1) event are actually sent
-        verify(message, times(times * (Integer.parseInt(sinkConnectorConfig.getMaxRetryTimes()) + 1))).create(any(), any());
+                Thread.sleep(duration);
```
[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/test/java/org/apache/eventmesh/connector/lark/sink/ImServiceHandlerTest.java:154]: At present, I have only simulated the scenario where the Lark server responds incorrectly, and used `verify` to verify that after passing the retry mechanism, the `create` method was indeed called the corresponding number of times (here it is 3*(3+1) times) . If there is a better way for testing, please give me some hints.

目前我只模拟了lark服务端响应错误的场景，并通过`verify`来验证通过重试机制后，确实调用了`create`方发相应的次数（在此是3*(3+1)次）。如果对于测试有更好办法的话，请给我一些提示。

File: eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/ImServiceHandler.java [hunk 9]
```diff
@@ -93,76 +102,154 @@ public static ImServiceHandler create(SinkConnectorConfig sinkConnectorConfig) {
...
                 log.warn("Sinking event to lark failure | code:[{}] | msg:[{}] | err:[{}]", resp.getCode(), resp.getMsg(), resp.getError());
                 return connectRecord;
             }
-
-            UN_ACK_REQ.remove(connectRecord);
             return null;
         });
     }
 
+    public void sinkAsync(ConnectRecord connectRecord) {
```
[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/ImServiceHandler.java:174]: I would like to know if there are any hidden dangers in  _submitting a task to the thread pool every time a message is sent_?

[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/ImServiceHandler.java:174]: ok,i will try to refactor it.

[the remark this answers is not in the document]
Comment (hhuang1231) [reply eventmesh-connectors/eventmesh-connector-lark/src/main/java/org/apache/eventmesh/connector/lark/sink/ImServiceHandler.java:174]: It's okay, I haven't started this work yet. Thank you for your attention to this PR, I have learned a lot.

没关系，我还没有开始这项工作。感谢您对这个pr的关注，让我学到了很多东西。

**Review by xwm1992 — APPROVED** [review]

**Pull request merged**